CRI: Support enable_unprivileged_icmp and enable_unprivileged_ports options

olljanat opened the pull request:
Borrowing logic from https://github.com/moby/moby/pull/41030 to solve #4936 (and https://github.com/kubernetes/kubernetes/issues/102612 ) by adding options `enable_unprivileged_icmp` and `enable_unprivileged_ports`. However in-this point those are disabled by default.

~Only open question is do we need similar logic like was added by https://github.com/moby/moby/pull/42736 here too? (I'm not familiar enough to be able answer that question).~

mikebrow reviewed (commented):
see comments.. General comment: I can't remember where we document the sysctl defaults but we should look around for that and update any of those locations.. and if we don't have a spot.. let's create one in the doc/cri directory pls... unless we add a disable option in cri config.. then you can just mention it where the new switch is added.

olljanat commented on pkg/cri/server/sandbox_run_linux.go:
```diff
@@ -132,8 +133,30 @@ func (c *criService) sandboxContainerSpec(id string, config *runtime.PodSandboxC
-	// Add sysctls
-	sysctls := config.GetLinux().GetSysctls()
+	// Add default sysctls that are generally safe and useful; currently we
+	// grant the capabilities to allow these anyway. You can override if
```
I mean like it is on last test case.  `PodSandboxConfig` can contain overwrite values.

mikebrow replied:
right .. the kubelet provided overwrite value...

olljanat commented on pkg/cri/opts/spec_linux.go:
```diff
@@ -593,6 +593,14 @@ func WithOOMScoreAdj(config *runtime.ContainerConfig, restrict bool) oci.SpecOpt
+// exist, so do not add the default ones if running on an old kernel.
+func SysctlExists(s string) bool {
+	f := filepath.Join("/proc", "sys", strings.Replace(s, ".", "/", -1))
+	_, err := os.Stat(f)
```
That is direct copy/paste from Docker code I assumed that it is well tested but I can add some debug logging here.

mikebrow replied:
the debug is for people to see that it's happening

olljanat commented on pkg/cri/server/sandbox_run_linux.go:
```diff
@@ -132,8 +133,30 @@ func (c *criService) sandboxContainerSpec(id string, config *runtime.PodSandboxC
+	sysctls := make(map[string]string)
+	if nsOptions.GetNetwork() != runtime.NamespaceMode_NODE {
+		// We cannot set up ping socket support in a user namespace
+		if !userns.RunningInUserNS() && customopts.SysctlExists("net.ipv4.ping_group_range") {
```
Not sure what you mean?

mikebrow replied:
example pattern here: https://github.com/containerd/containerd/blob/main/pkg/cri/opts/spec_linux.go#L503-L518

olljanat replied:
Ah ok. Anyway only purpose of `SysctlExists` function was to check if those sysctls exists as original proposal was enable these automatically without disable switch. But as we changed plan that part does not exist all anymore.

cpuguy83 commented:
This should be exposed as an option.
This change did cause some regressions for folks in docker.

olljanat commented:
@cpuguy83 would be interesting to know about which kind of regressions it caused on Docker? I can see from cosebase that user namespace detection for ping was incorrect on original PR and was fixed later on but that I already copied to here.

If decission is that this need to be done as option then probably it would be better to use same approach than CRI-O and add `default_sysctls=[]` to config file?

cpuguy83 commented:
@olljanat I spent a little time trying to find them... but it's a big haystack. Maybe @thaJeztah has some indexed.

thaJeztah commented:
Thanks for the ping; I had actually bookmarked this PR, because I wanted to comment if this (for containerd) should be the default; docker is somewhat more opinionated in its config, but perhaps for containerd (intended to be a "lower level" components to be built upon), it should indeed be an option.

But then I saw this change was part of the CRI code, so perhaps that code is "more opinionated"?

> I spent a little time trying to find them... but it's a big haystack. Maybe @thaJeztah has some indexed.

So there's some corner-cases, and some cases that need to be looked into; I can find links, but from the top of my head;

- container running with `--network=host`
- container running on (in docker) a MacVLAN or IPVlan network (so container is part of an externally managed network)

In both of the above, changing this setting could be somewhat surprising (as now the config doesn't affect the "container network sandbox", but actually the network in which the container is active.

The more "corner case" scenario was for a regular container (with port mapping), but the container was used as an environment to run unit/integration tests, with the assumption that "container environment" would act the same as a "host environment" (so "privileged ports" to be blocked). I don't think that scenario was really critical (and easy to work around by overriding the configs to use its details), but worth mentioning.

olljanat commented:
@thaJeztah good points. Even if we add setting to enable/disable this logic those special cases most likely need to be handled somehow. At least the most common ones. Host network is already handled but now I wonder that if there is some CNI plugin which with we might see issues after this is enabled?

thaJeztah commented:
> now I wonder that if there is some CNI plugin which with we might see issues after this is enabled?

I'll defer that question to others, as I don't have a lot of experience in that area (on k8s) 😅, but hope that my examples help identifying _possible_ situations where this setting could be problematic / undesirable (or perhaps not at all!)

olljanat commented:
Modified to working on way that there is `enable_unprivileged_icmp` and `enable_unprivileged_ports` settings which are not enabled by default. Also dropped SysctlExists logic as we can assume that user know what he is doing when enabling those.

Added as separate commit for now. Will combine those after we are able to agree if this is correct approach (and will fix those variable names which CI is complaining then).

mikebrow commented:
/ok-to-test

mikebrow commented:
> This should be exposed as an option.
> This change did cause some regressions for folks in docker.

that's what I was thinking to.. probably needs an option or two for the cri config

mikebrow commented on docs/cri/config.md:
```diff
@@ -107,6 +107,16 @@ version = 2
+  # enable_unprivileged_icmp configures net.ipv4.ping_group_range="0 2147483647"
+  # for all containers which are not using host network, are not running in user namespace
+  # and if it is not overwritten by PodSandboxConfig
+  enable_unprivileged_icmp = true
```
which version of kubernetes will this be expected to be set to true on? should probably mention that here and in the config.go..

mikebrow replied:
if you want the default to be true you need to set it in config_unix.go  you'll see the defaults there..

if the expectation is this will default on in this upcoming version of kubernetes.. I suggest making this a disable boolean with a default of false..

mikebrow replied:
if this is more alpha.. then stick with enable... and link to the feature in k8s where this will become a desired default

and stick with enable default false.. till r.next or at least a service refresh and we can change the default to true in config_unix.go when the stars align with the kubelet changes

olljanat replied:
I understood from comments that we don't want enable these by default here because we cannot be sure that it does not break some existing use cases/configurations. That why it is now on purpose in way that those settings are not enabled by default but documentation still suggests to enabling them by showing that on examples.

olljanat replied:
Ah, just noticed that description text says `The explanation and default value of each configuration item are as follows:` so will change those to `false`

olljanat replied:
> if this is more alpha.. then stick with enable... and link to the feature in k8s where this will become a desired default

Not sure where you want to link? Add comment to code which links to K8s github?

mikebrow replied:
was hoping for some reference to the discussion for context to the reader ```[k8s discussion](https://github.com/kubernetes/kubernetes/issues/102612)```

mikebrow replied:
@crosbymichael note the.. default off for now.. response :-(

olljanat replied:
Link added now.

mikebrow reviewed (commented):
see comments

olljanat commented:
Fixed documentation matching to default settings and squashed to single commit. Also one things which just came to my mind as now this does not change any existing logic is that can we actually get it backported to 1.5.x versions too? Would be nice to be able use it already on those K3s/RKE2 v1.22.x versions.

olljanat commented:
@mikebrow I hopefully covered all your comments on latest update. CI failure looks like infra issue for me.

olljanat commented:
@mikebrow there looks to be some odd issue on CI servers but does this other way now looks like you wanted it to be?

BenTheElder commented:
Did I read correctly that this is enabled by default in docker still but not here?

Can we align this on default-enable with the option to disable? (Perhaps 1.5.X backport if any would disable by default though)
Or does Docker plan to revert this default?

cpuguy83 reviewed (approved):
LGTM

cpuguy83 commented:
@BenTheElder This change can cause some undesirable and breaking behavior. I would prefer not to enable this by default.

Worth noting on this PR the KEP for ambient caps: https://github.com/kubernetes/enhancements/pull/2757

BenTheElder commented:
That doesn't answer my questions though.

When cri-containerd and dockershim deviate on things like this, users get very frustrated that a deployment works with one and not the other.

See e.g. this one https://github.com/kubernetes-sigs/kind/issues/1331

crosbymichael commented:
@BenTheElder I read over the k8s issues and I didn't see consensus yet where the suggestion is for this to be enabled by default. Am I missing anything?

BenTheElder commented:
Unless I've misunderstood this *is* enabled by default in docker, and therefore with dockershim users?

On the k8s side of things I only see comments suggesting it should be enabled by default (e.g. https://github.com/kubernetes/kubernetes/issues/102612#issue-911968751 https://github.com/kubernetes/kubernetes/issues/56374#issuecomment-868892588).

cpuguy83 commented:
@BenTheElder What we end up with in this case is potentially breaking existing user's production deployments.
It might be worth doing in a 2.0 or something, but I'm worried to do it now.

dmcgowan commented:
@cpuguy83 agreed, it is worth opening up such an issue and we can milestone it for 2.0. We are starting to plan out what items we might want to consider for deprecation or breaking change in a 2.0. In terms of consistently, it is more important for us to stay consistent with upgrading containerd 1.x users than users moving from another runtime. We have never guaranteed the default behavior will match Docker, however, we should document that better. We have talked about having a migration guide but still don't have one today.

thaJeztah commented:
Agreed; also (per my earlier comment); Docker may have options that are more "opinionated" than containerd (and containerd cri), which may have more "tweakability" / require more configuration, so I don't think they necessarily have to match.

jonyhy96 commented on pkg/cri/server/sandbox_run_linux.go:
```diff
@@ -134,6 +135,19 @@ func (c *criService) sandboxContainerSpec(id string, config *runtime.PodSandboxC
+			sysctls["net.ipv4.ip_unprivileged_port_start"] = "0"
+		}
+		if c.config.EnableUnprivilegedICMP && !pingGroupRange && !userns.RunningInUserNS() {
+			sysctls["net.ipv4.ping_group_range"] = "0 2147483647"
```
@olljanat can u explain why use 2147483647 over 4294967295

olljanat replied:
@jonyhy96 like first message says these are borrowed directly from Docker engine code: https://github.com/moby/moby/blob/885f3c65d54f7c85461591cda8bd9dcfb5a9723f/daemon/oci_linux.go#L775-L776

jonyhy96 replied:
Do u know the reason why they use that as range？

olljanat replied:
@justincormack can you comment about this on?

justincormack replied:
from https://systemd.io/UIDS-GIDS/

> Note that the range 2147483648…4294967294 (i.e. 2^31…2^32-2) should be handled with care. Various programs (including kernel file systems, see devpts) have trouble with UIDs outside of the signed 32bit range, i.e any UIDs equal to or above 2147483648. It is thus strongly recommended to stay away from this range in order to avoid complications. This range should be considered reserved for future, special purposes.

In general I think the "negative" (yes its unsigned in theory but...) uids and gids are basically unusable.

justincormack replied:
also 4294967295 is defintiely unusable as it is the failure return value and placeholder for dont set on many syscalls.

brandond commented:
@cpuguy83 can you explain more how enabling these breaks existing deployments? There's been an ask to enable these by default for K3s/RKE2 and I'd like to understand better what the risk is.
